devicetypes/smartthings/zigbee-valve.src/zigbee-valve.groovy

```diff
@@ -11,100 +11,138 @@
+            state "default", label:"", action:"refresh.refresh", icon:"st.secondary.refresh"
+        }
+
+        main(["valve"])
+        details(["valve", "battery", "refresh"])
+    }
 }
 
-def off() {
-	log.debug "off()"
-	sendEvent(name: "switch", value: "off")
-	"st cmd 0x${device.deviceNetworkId} 1 6 0 {}"
+private getCLUSTER_BASIC() { 0x0000 }
+private getBASIC_ATTR_POWER_SOURCE() { 0x0007 }
+private getCLUSTER_POWER() { 0x0001 }
+private getPOWER_ATTR_BATTERY_PERCENTAGE_REMAINING() { 0x0021 }
+private getTYPE_U8() { 0x20 }
+private getTYPE_ENUM8() { 0x30 }
+
+// Parse incoming device messages to generate events
+def parse(String description) {
+    log.debug "description is $description"
+    def event = zigbee.getEvent(description)
+    if (event) {
+        if(event.name == "switch") {
+            event.name = "contact"			//0006 cluster in valve is tied to contact
+            if(event.value == "on") {
+                event.value = "open"
+            }
+            else if(event.value == "off") {
+                event.value = "closed"
+            }
+            sendEvent(event)
+        }
+        else {
+            sendEvent(event)
+        }
+    }
+    else {
+        def descMap = zigbee.parseDescriptionAsMap(description)
+        if (descMap?.clusterInt == CLUSTER_BASIC && descMap?.attrInt == BASIC_ATTR_POWER_SOURCE){
+            def value = descMap?.value
+            if (value == "01" || value == "02") {
+                sendEvent(name: "powerType", value: "Mains")
+            }
+            else if (value == "03") {
+                sendNotification("$linkText valve is now running on reserve battery power")
+                sendEvent(name: "powerType", value: "Battery")
+            }
+            else if (value == "04") {
+                sendEvent(name: "powerType", value: "DC")
+            }
+            else {
+                sendEvent(name: "powerType", value: "Unknown")
+            }
+        }
+        else if (descMap.clusterInt == CLUSTER_POWER && descMap.attrInt == POWER_ATTR_BATTERY_PERCENTAGE_REMAINING) {
+            event.name = "battery"
+            event.value = Math.round(Integer.parseInt(descMap.value, 16) / 2)
+            sendEvent(event)
+        }
+        else {
+            log.warn "DID NOT PARSE MESSAGE for description : $description"
+            log.debug descMap
+        }
+    }
 }
 
 def open() {
-	log.debug "on()"
-	sendEvent(name: "switch", value: "on")
-	"st cmd 0x${device.deviceNetworkId} 1 6 1 {}"
+    zigbee.on()
 }
 
 def close() {
-	log.debug "off()"
-	sendEvent(name: "switch", value: "off")
-	"st cmd 0x${device.deviceNetworkId} 1 6 0 {}"
+    zigbee.off()
 }
 
 def refresh() {
-	log.debug "sending refresh command"
-	"st rattr 0x${device.deviceNetworkId} 1 6 0"
+    log.debug "refresh called"
+    zigbee.onOffRefresh() +
+    zigbee.readAttribute("${CLUSTER_BASIC}", "${BASIC_ATTR_POWER_SOURCE}") +
```

Comment:
These next few lines should have the `${..}` removed. For example, this line should be simply:

`zigbee.readAttribute(CLUSTER_BASIC, BASIC_ATTR_POWER_SOURCE)`